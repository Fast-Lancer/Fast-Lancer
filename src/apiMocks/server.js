import { rest } from 'msw'
import { setupServer } from 'msw/node'

const url = process.env.REACT_APP_SUPABASE_URL + '/rest/v1'

let postedClient = {}
let postedProject = {}

export const server = setupServer(
  rest.get(url + '/clients', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(select === '*' && !id) {
      // response for getClients()
      return res(
        ctx.json([
          {
            id: 1,
            client_name: 'bob1',
            business_name: 'business1'
          },
          {
            id: 2,
            client_name: 'bob2',
            business_name: 'business2'
          },
          {
            id: 3,
            client_name: 'bob3',
            business_name: 'business3'
          }
        ])
      )
    } else if(id) {
      // response for getClient(id)
      if(id === 'eq.42') {
        // Janky workaround to make testing edit/create possible.
        // The post route mock will set the id to 42
        // So this responds with the posted data when client.id is 42
        return res(
          ctx.json(
            postedClient
          )
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              client_name: 'bob1',
              business_name: 'business1'
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/clients', (req, res, ctx) => {
    let [client] = req.body
    client = { ...client, id: 42 }
    postedClient = client
    return res(
      ctx.json([
        client
      ])
    )
  }),
  rest.get(url + '/projects', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(!id) {
      // response for getProjects()
      return res(
        ctx.json([
          {
            id: 1,
            title: 'project1',
            client_id: '1',
            clients: {
              client_name: 'bob1'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          },
          {
            id: 2,
            title: 'project2',
            client_id: '2',
            clients: {
              client_name: 'bob2'
            },
            
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          
          },
          {
            id: 3,
            title: 'project3',
            client_id: '3',
            clients: {
              client_name: 'bob3'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          }
        ])
      )
    } else if(id) {
      // response for getProjectById(id)
      if(id === 'eq.42') {
        return res(
          ctx.json(postedProject)
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              title: 'project1',
              client_id: '1',
              clients: {
                client_name: 'bob1'
              },
              date_start:'2022-01-20',
              date_end:'2022-01-21',
              URL: 'something@something.com',
              description: 'ok',
              notes: 'test notes',
              hourly_rate: 33,
              hours_quoted: 22,
              price_quoted: 300
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/projects', (req, res, ctx) => {
    let project = req.body
    project = { ...project, id: 42 }
    postedProject = project
    return res(
      ctx.json([project])
    )
  })
)
