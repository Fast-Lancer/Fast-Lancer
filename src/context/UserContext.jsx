import { createContext, useContext, useMemo, useState } from 'react'
import { getUser } from '../services/users'

const UserContext = createContext({})

export function UserProvider({ children }) {
  const currentUser = getUser()
  const [user, setUser] = useState(currentUser ?? {})
  const memo = useMemo(() => ({ user, setUser }), [user])
  return (
    <UserContext.Provider value={memo}>
      {children}
    </UserContext.Provider>
  )
}

export function useUser() {
  const context = useContext(UserContext)

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider')
  }

  return context
};
