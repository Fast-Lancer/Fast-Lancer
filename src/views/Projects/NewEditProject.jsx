import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { useHistory, useLocation } from 'react-router-dom'
import Button from '../../components/Button/Button.jsx'
import ProjectForm from '../../components/ProjectForm/ProjectForm.jsx'
import { getClients } from '../../services/clients.js'
import { createProject, deleteProjectById, getProjectNoClient, updateProject } from '../../services/projects.js'
import changeValue from '../../utils/changeValue.js'

export default function NewEditProject() {
  const history = useHistory()
  const location = useLocation()
  const currentLocation = location.pathname
  const isCreate = currentLocation === '/projects/new'
  const [initialValues, setInitialValues] = useState({
    date_start: '', date_end: '', hourly_rate: '', hours_worked: '', URL: '', description: '', completed_at: '', price_quoted: '', hours_quoted: '', title: '', notes: '', client_id: '', user_id: ''
  })

  const [loading, setLoading] = useState(true)
  const [clientsAvailable, setClientsAvailable] = useState([{ client_name: '', id: '' }])

  const { id } = useParams()

  const handleProject = async (form) => {
    try{
      let project
      if(isCreate) {
        project = await createProject(form)
      } else {
        project = await updateProject(form)
      }

      history.push(isCreate ? `/projects/${project[0].id}` : `/projects/${form.id}`)

    } catch(e) {
    }
  }

  const handleClick = () => {
    // setTimeout so /projects reloads with updated information after delete
    deleteProjectById(id).then(() => setTimeout(() => {history.replace('/projects')}, 100))
  }

  useEffect(() => {
    if (!isCreate) {
      getProjectNoClient(id).then(proj => changeValue(proj, '')).then(setInitialValues)
    }
    getClients().then(setClientsAvailable).finally(() => setLoading(false))
  }, [])

  return <div>
    {loading
      ? <h1>Loading...</h1>
      : <ProjectForm {...{ handleProject, clientsAvailable, initialValues }}/>
    }
    {!isCreate && <Button handleClick={handleClick} buttonText={'Delete'}/>}
  </div>
}
