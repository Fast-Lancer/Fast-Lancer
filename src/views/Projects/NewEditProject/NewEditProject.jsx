import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { useHistory, useLocation } from 'react-router-dom'
import ProjectForm from '../../../components/ProjectForm/ProjectForm.jsx'
import Title from '../../../components/Title/Title.jsx'
import { getClients } from '../../../services/clients.js'
import { createProject, deleteProjectById, getProjectById,  updateProject } from '../../../services/projects.js'
import changeValue from '../../../utils/changeValue.js'

export default function NewEditProject() {
  const history = useHistory()
  const location = useLocation()
  const currentLocation = location.pathname
  const isCreate = currentLocation === '/projects/new'
  const [initialValues, setInitialValues] = useState({
    date_start: '', date_end: '', hourly_rate: '', hours_worked: '', URL: '', description: '', completed_at: '', price_quoted: '', hours_quoted: '', title: '', notes: '', client_id: '', user_id: ''
  })
  const title = isCreate ? 'New Project' : 'Edit Project'

  const [loading, setLoading] = useState(true)
  const [clientsAvailable, setClientsAvailable] = useState([{ client_name: '', id: '' }])

  const { id } = useParams()

  const handleProject = async (form) => {
    try{
      const project = await(isCreate ? createProject(form) : updateProject(form))

      history.push(isCreate ? `/projects/${project[0].id}` : `/projects/${form.id}`)

    } catch(err) {
      throw new Error(err)
    }
  }

  const handleDelete = () => {
    // setTimeout so /projects reloads with updated information after delete
    deleteProjectById(id).then(() => setTimeout(() => {history.replace('/projects')}, 100))
  }


  useEffect(() => {
    if (!isCreate) {
      getProjectById(id).then(proj => changeValue(proj, '')).then(setInitialValues).then(() => getClients()).then(setClientsAvailable).finally(() => setLoading(false))
    } else {
      getClients().then(setClientsAvailable).finally(() => setLoading(false))
    }
  }, [])

  return <div>
    <Title pageTitle={'new-edit project'} pageHeader={title}/> 
    {loading
      ? <h1>Loading...</h1>
      : <ProjectForm {...{ handleProject, clientsAvailable, initialValues, isCreate, handleDelete }}/>
    }
  </div>
}
