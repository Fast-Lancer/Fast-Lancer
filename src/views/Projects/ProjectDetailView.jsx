import { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { useParams } from 'react-router-dom'
import Button from '../../components/Button/Button.jsx'
import ProjectDetail from '../../components/ProjectDetail/ProjectDetail.jsx'
import Title from '../../components/Title/Title.jsx'
import { getProjectById } from '../../services/projects.js'

export default function ProjectDetailView() {
  const { id } = useParams()
  const [project, setProject] = useState({})
  const [loading, setLoading] = useState(true)
  
  useEffect(() => {
    getProjectById(id).then(data => setProject(data)).finally(() => setLoading(false))
  }, [])

  if(loading) return <h1>Loading...</h1>
  return (
    <div>
      <Title pageTitle='project detail' pageHeader={project.title} />
      <ProjectDetail project={project} />
    </div>
  )
}
