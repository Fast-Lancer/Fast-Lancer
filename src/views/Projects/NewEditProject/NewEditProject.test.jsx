import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { useEffect } from 'react'
import { Route, Switch } from 'react-router-dom'
import { useLocation } from 'react-router-dom'
import { MemoryRouter } from 'react-router-dom'
import { server } from '../../../apiMocks/server.js'
import { UserProvider, useUser } from '../../../context/UserContext.jsx'
import ProjectDetailView from '../ProjectDetailView/ProjectDetailView.jsx'
import NewEditProject from './NewEditProject.jsx'

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})

const UserAdder = () => {
  const { setUser } = useUser()
  useEffect(() => {
    setUser({ id: 1 })
  }, [])

  return null
}

let location

const LocationGrabber = () => {
  location = useLocation()
  return null
}

it.skip('should render the fetched data', async () => {
  const { container } = render(
    <MemoryRouter initialEntries={['/projects/new']}>
      <UserProvider>
        <UserAdder />
        <Switch>
          <Route path='/projects/new'>
            <LocationGrabber />
            <NewEditProject />
          </Route>
          <Route path='/projects/:id'>
            <LocationGrabber />
            <ProjectDetailView />
          </Route>
        </Switch>
      </UserProvider>
    </MemoryRouter>
  )

  const sDateInput = await screen.findByLabelText(/Start Date:/)
  const eDateInput = await screen.findByLabelText(/End Date:/)
  const rateInput = await screen.findByLabelText(/Hourly Rate:/)
  const workedInput = await screen.findByLabelText(/Hours Worked:/)
  const urlInput = await screen.findByLabelText(/URL:/)
  const descInput = await screen.findByLabelText(/Description:/)
  const completedInput = await screen.findByLabelText(/Completed At:/)
  const pQuoteInput = await screen.findByLabelText(/Price Quoted:/)
  const hQuoteInput = await screen.findByLabelText(/Hours Quoted:/)
  const titleInput = await screen.findByLabelText(/Title:/)
  const notesInput = await screen.findByLabelText(/Notes:/)
  const clientInput = await screen.findByLabelText(/Client:/)
  const submitButton = await screen.findByText('Save')

  fireEvent.change(sDateInput, { target: { value: '1-1-2000' } })
  fireEvent.change(eDateInput, { target: { value: '1-1-2000' } })
  fireEvent.change(rateInput, { target: { value: '2' } })
  fireEvent.change(workedInput, { target: { value: '3' } })
  fireEvent.change(urlInput, { target: { value: 'abc.com' } })
  fireEvent.change(descInput, { target: { value: 'desc' } })
  fireEvent.change(completedInput, { target: { value: '1-1-2000' } })
  fireEvent.change(pQuoteInput, { target: { value: '4' } })
  fireEvent.change(hQuoteInput, { target: { value: '5' } })
  fireEvent.change(titleInput, { target: { value: 'test title' } })
  fireEvent.change(notesInput, { target: { value: 'test notes' } })
  fireEvent.change(clientInput, { target: { value: 1 } })

  fireEvent.click(submitButton)

  await waitFor(() => location.pathName === '/projects/42', { timeout: 1000 })

  await screen.findByText('test title')
  await screen.findByText('test notes')

  expect(container).toMatchSnapshot()
})
