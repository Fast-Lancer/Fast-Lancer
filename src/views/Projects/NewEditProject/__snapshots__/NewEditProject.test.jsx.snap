// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the fetched data 1`] = `
<div>
  <div>
    <div
      class="titleBody"
    >
      <ul
        class="tabnav"
      >
        <li
          class="active"
        >
          <a
            href="/projects"
          >
            Projects
          </a>
        </li>
        <li>
          <a
            href="/clients"
          >
            Clients
          </a>
        </li>
        <li>
          <a
            href="/projects/edit/42"
          >
            Edit Project
          </a>
        </li>
        <li>
          <a
            href="/login"
          >
            Log Out
          </a>
        </li>
      </ul>
    </div>
    <main>
      <div
        class="projectDetail"
      >
        <div
          class="projectTitle"
        >
          test title
        </div>
        <div
          class="clientName"
        >
          <a
            href="/clients/1"
          >
            <h1
              aria-label="client-name"
            >
              No client name entered
            </h1>
          </a>
          <h3>
            Client
          </h3>
        </div>
        <div
          class="infoWrapper"
        >
          <h3>
            <div
              aria-label="URL"
              class="infoItem"
            >
              <a
                href="/projects/abc.com"
              >
                <img
                  alt="www"
                  src="wwwSmall.png"
                />
                abc.com
              </a>
            </div>
          </h3>
          <section
            class="infoSection"
          >
            <div
              aria-label="hourly-rate"
              class="infoItem"
            >
              <img
                alt="money"
                src="moneySmall.png"
              />
              Hourly Rate: $
              2
            </div>
            <div
              aria-label="start-date"
              class="infoItem"
            >
              <img
                alt="clock"
                src="clockSmall.png"
              />
              Started: 
            </div>
            <div
              aria-label="hours-logged"
              class="infoItem"
            >
              <img
                alt="money"
                src="moneySmall.png"
              />
              Hours Logged: 
              3
            </div>
            <div
              aria-label="end-date"
              class="infoItem"
            >
              <img
                alt="clock"
                src="clockSmall.png"
              />
              Deadline: 
            </div>
          </section>
        </div>
        <section
          class="alignLeft"
        >
          <h2>
            Description
          </h2>
          <p
            aria-label="description"
            class="textBlock"
          >
            desc
          </p>
          <h2>
            Notes
          </h2>
          <p
            aria-label="notes"
            class="textBlock"
          >
            test notes
          </p>
        </section>
      </div>
    </main>
  </div>
</div>
`;
